pkg/cri/server/container_create.go

```diff
@@ -101,6 +101,10 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to get image from containerd %q", image.ID)
 	}
+	imageConfig := imagespec.ImageConfig{}
```

Comment:
I think we can use brand new env string slice instead of json.Marshal and Unmarshal.

```golang
env := append([]string{}, imageConfig.Env...)
```

Marshal/Unmarshal is heavy call than `append([]string{}, ...)`

Reply:
Yes, It is heavy but is more secure to avoid some hidden bugs like this issue which is very hard to reproduce and debug.

Reply:
Well, we can unit test to cover the case, or use assign fields to handle deep copy instead marshal/unmarshal. My only concern is that cost from marshal/unmarshal.

Reply:
I like @fuweid 's idea to change 
```
	// Apply envs from image config first, so that envs from container config
	// can override them.
	env := imageConfig.Env
	for _, e := range config.GetEnvs() {
		env = append(env, e.GetKey()+"="+e.GetValue())
	}
```
to 
```
	// Apply envs from image config first, so that envs from container config
	// can override them.
	env := append([]string{}, config.Env...)
	for _, e := range config.GetEnvs() {
		env = append(env, e.GetKey()+"="+e.GetValue())
	}
```
and after talking to fuwei maybe we can do the copy down under the resolve code in / Get gets image metadata by image id. The id can be truncated.
// Returns various validation errors if the image id is invalid.
// Returns storeutil.ErrNotExist if the image doesn't exist.
func (s *Store) Get(id string) (Image, error) {
	return s.store.get(id)
}
@dmcgowan

Reply:
historical side note: for this cri code the image config was originally a copy because we received it over grpc (when cri was an external daemon)